File: iOS/Sources/Data/Repositories/InMemoryExpenseRepository.swift
import Foundation

class InMemoryExpenseRepository: ExpenseRepositoryProtocol {
    private var expenses: [Expense] = []
    private var nextId: Int64 = 1

    func fetchAll() async throws -> [Expense] {
        return expenses.sorted { $0.date > $1.date }
    }

    func fetch(id: Int64) async throws -> Expense {
        guard let expense = expenses.first(where: { $0.id == id }) else {
            throw DomainError.notFound
        }
        return expense
    }

    func fetchByBudgetId(_ budgetId: Int64) async throws -> [Expense] {
        return expenses
            .filter { $0.budgetId == budgetId }
            .sorted { $0.date > $1.date }
    }

    func fetchByDateRange(start: Date, end: Date) async throws -> [Expense] {
        return expenses
            .filter { $0.date >= start && $0.date <= end }
            .sorted { $0.date > $1.date }
    }

    func add(_ expense: Expense) async throws -> Expense {
        var newExpense = expense
        newExpense = Expense(
            id: nextId,
            budgetId: expense.budgetId,
            amount: expense.amount,
            description: expense.description,
            date: expense.date,
            createdAt: expense.createdAt,
            updatedAt: expense.updatedAt
        )
        nextId += 1
        expenses.append(newExpense)
        return newExpense
    }

    func update(_ expense: Expense) async throws {
        guard let index = expenses.firstIndex(where: { $0.id == expense.id }) else {
            throw DomainError.notFound
        }
        var updated = expense
        updated = Expense(
            id: expense.id,
            budgetId: expense.budgetId,
            amount: expense.amount,
            description: expense.description,
            date: expense.date,
            createdAt: expense.createdAt,
            updatedAt: Date()
        )
        expenses[index] = updated
    }

    func delete(id: Int64) async throws {
        guard expenses.contains(where: { $0.id == id }) else {
            throw DomainError.notFound
        }
        expenses.removeAll { $0.id == id }
    }

    func hasExpenses(forBudgetId budgetId: Int64) -> Bool {
        expenses.contains { $0.budgetId == budgetId }
    }
}

File: android/app/src/main/java/com/budgetmanager/data/repository/InMemoryExpenseRepository.kt
package com.budgetmanager.data.repository

import com.budgetmanager.domain.entity.DomainError
import com.budgetmanager.domain.entity.Expense
import com.budgetmanager.domain.repository.ExpenseRepository
import java.time.LocalDate
import java.time.LocalDateTime

class InMemoryExpenseRepository : ExpenseRepository {
    private val expenses = mutableListOf<Expense>()
    private var nextId: Long = 1

    override suspend fun getAll(): List<Expense> {
        return expenses.sortedByDescending { it.date }
    }

    override suspend fun getById(id: Long): Expense {
        return expenses.find { it.id == id } ?: throw DomainError.NotFound
    }

    override suspend fun getByBudgetId(budgetId: Long): List<Expense> {
        return expenses
            .filter { it.budgetId == budgetId }
            .sortedByDescending { it.date }
    }

    override suspend fun getByDateRange(startDate: LocalDate, endDate: LocalDate): List<Expense> {
        return expenses
            .filter { it.date in startDate..endDate }
            .sortedByDescending { it.date }
    }

    override suspend fun add(expense: Expense): Expense {
        val newExpense = expense.copy(id = nextId++)
        expenses.add(newExpense)
        return newExpense
    }

    override suspend fun update(expense: Expense) {
        val index = expenses.indexOfFirst { it.id == expense.id }
        if (index == -1) throw DomainError.NotFound

        expenses[index] = expense.copy(updatedAt = LocalDateTime.now())
    }

    override suspend fun delete(id: Long) {
        if (!expenses.removeIf { it.id == id }) {
            throw DomainError.NotFound
        }
    }

    override fun hasExpenses(budgetId: Long): Boolean {
        return expenses.any { it.budgetId == budgetId }
    }
}
